File: src/my_robot_tutorials/src/number_counter.cpp
#include <ros/ros.h>
#include <std_msgs/Int64.h>
#include <std_srvs/SetBool.h>


int global_counter = 0;
ros::Publisher pub;


void callback_receive_number(const std_msgs::Int64& msg)
{
    global_counter++;
    ROS_INFO("Message Received: %ld", msg.data);
    std_msgs::Int64 count_msg;        
    count_msg.data = global_counter;
    pub.publish(count_msg);
}


bool handle_reset_counter(
    std_srvs::SetBool::Request &req,
    std_srvs::SetBool::Response &res
)
{
    if(req.data) {
        ROS_WARN("Service called, Resetting counter to 0");
        global_counter = 0;
        res.success = true;
        res.message = "Counter was reset";
        return true;
    }

    ROS_WARN("Service called, But not Resetting counter");
    res.success = true;
    res.message = "Counter was not reset";
    return true;
}


int main(int argc, char **argv) 
{
    ros::init(argc, argv, "smartphone_node");
    ros::NodeHandle nh;

    ros::Subscriber sub = nh.subscribe("/number", 1000, callback_receive_number);
    pub = nh.advertise<std_msgs::Int64>("/number_count", 10);

    ros::ServiceServer server = nh.advertiseService(
        "/reset_counter",
        handle_reset_counter
    );

    // Keep the node running until we send a shutdown signal to it
    ros::spin();
}

File: src/my_robot_tutorials/scripts/number_counter.py
#!/usr/bin/env python3

import rospy
from std_msgs.msg import Int64
from std_srvs.srv import SetBool


num_counter = 0
pub = None

def callback_receive_number(msg):
    global num_counter
    rospy.loginfo("Message received: " + str(msg.data))
    num_counter += msg.data
    new_msg = Int64()
    new_msg.data = num_counter
    pub.publish(new_msg)

def handle_reset_counter(req):
    global num_counter
    if req.data:
        num_counter = 0    
        rospy.logwarn("Counter was reset")
        return True, "Counter was resetted after service call"
    rospy.logwarn("Service called but counter was not reset")
    return True, "Counter was not reset since the request data was false"


if __name__ == "__main__":
    # anonymous will add a random number after the name of the node. Therefore we can run multiple 
    # instances of this node
    rospy.init_node("number_counter", anonymous=True)

    sub = rospy.Subscriber("/number", Int64, callback_receive_number)
    pub = rospy.Publisher("/number_count", Int64, queue_size=10)

    service = rospy.Service(
        "/reset_counter",  # Name of service
        SetBool,  # Type of request
        handle_reset_counter  # Handle function
    )
    rospy.loginfo("Service Server has been created")

    # Keep the script here without going further
    rospy.spin()
